### PR TITLE
Reducing string allocation in Base64UrlTextEncoder's Encode path

### PR DESCRIPTION
**Pull request by kichalla**

### CODE REVIEW

File: src/Microsoft.AspNetCore.WebUtilities/Base64UrlTextEncoder.cs [hunk 1]
```diff
@@ -32,38 +32,30 @@ public static byte[] Decode(string text)
...
+                    {
+                        destination[i] = '-';
+                    }
+                    else if (destination[i] == '/')
+                    {
+                        destination[i] = '_';
+                    }
+                    else if (destination[i] == '=')
+                    {
+                        destination[i] = '.';
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.WebUtilities/Base64UrlTextEncoder.cs]: Why this change?

File: test/Microsoft.AspNetCore.WebUtilities.Tests/Base64UrlTextEncoderTests.cs [hunk 2]
```diff
@@ -65,5 +70,26 @@ public void DecodeToBase64String_ReturnsValid_Base64String(string text, string e
...
+                {
+                    data[index] = (byte)(5 + length + (index * 23));
+                }
+                string text = Base64UrlTextEncoder.Encode(data);
+
+                // Act
+                var urlEncodedText = UrlEncoder.Default.Encode(text);
+
+                // Assert
+                Assert.Equal(text, urlEncodedText);
```
Comment (cesarbs) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/Base64UrlTextEncoderTests.cs]: `Assert.NotSame`?

File: src/Microsoft.AspNetCore.WebUtilities/Base64UrlTextEncoder.cs (hunk 1 above)
Comment (kichalla) [inline src/Microsoft.AspNetCore.WebUtilities/Base64UrlTextEncoder.cs]: This is because `=` character gets url encoded and also if we were to have similar behavior like earlier we would need to allocate a new string (like InplaceStringBuilder did earlier internally)

File: test/Microsoft.AspNetCore.WebUtilities.Tests/Base64UrlTextEncoderTests.cs (hunk 2 above)
Comment (kichalla) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/Base64UrlTextEncoderTests.cs]: Do we care here? UrlEncoder might choose to do no-op since it couldn't find any characters to encode. We only want to make sure that characters remain the same.

Comment (cesarbs) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/Base64UrlTextEncoderTests.cs]: I see.

**Comment by Tratcher** [comment]
Perf data?

File: src/Microsoft.AspNetCore.WebUtilities/Base64UrlTextEncoder.cs [hunk 3]
```diff
@@ -32,38 +32,30 @@ public static byte[] Decode(string text)
             return Convert.FromBase64String(DecodeToBase64String(text));
         }
 
+        // Since the string instnace passed here is owned by the Encode method, its safe to modify
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.WebUtilities/Base64UrlTextEncoder.cs]: nit: instance

File: test/Microsoft.AspNetCore.WebUtilities.Tests/Base64UrlTextEncoderTests.cs [hunk 4]
```diff
@@ -32,30 +33,34 @@ public void DataOfVariousLengthRoundTripCorrectly()
         [InlineData("", "")]
         [InlineData("+", "-")]
         [InlineData("/", "_")]
-        [InlineData("=", "")]
-        [InlineData("==", "")]
-        [InlineData("a+b+c+==", "a-b-c-")]
-        [InlineData("a/b/c==", "a_b_c")]
-        [InlineData("a+b/c==", "a-b_c")]
+        [InlineData("=", ".")]
+        [InlineData("==", "..")]
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/Base64UrlTextEncoderTests.cs]: This may be problematic, .. has special meaning in url paths.

Comment (Tratcher) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/Base64UrlTextEncoderTests.cs]: so does `.` actually

File: src/Microsoft.AspNetCore.WebUtilities/Base64UrlTextEncoder.cs [hunk 5]
```diff
@@ -87,17 +78,16 @@ internal static string DecodeToBase64String(string text)
                 {
                     inplaceStringBuilder.Append('/');
                 }
+                else if (text[i] == '.')
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.WebUtilities/Base64UrlTextEncoder.cs]: You would only want to attempt this at the end of the string.

File: test/Microsoft.AspNetCore.WebUtilities.Tests/Base64UrlTextEncoderTests.cs (hunk 4 above)
Comment (Tratcher) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/Base64UrlTextEncoderTests.cs]: let's find a different replacement char

Comment (kichalla) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/Base64UrlTextEncoderTests.cs]: Ah OK

**Comment by kichalla** [comment]
Discussed offline with @Tratcher and @rynowak. Closing this PR as it's not worth the risk we might be taking.
